NIFI-6539 Nifi stateless init from flow xml

dbkegley opened the pull request:
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

nifi-stateless currently supports running a flow from a registry given a flow_id and a registry endpoint.  To facilitate environments where there is no connectivity to a registry, I would like the ability to run nifi-stateless from a flow.xml.gz on the filesystem.  When running in a container the flow.xml.gz could be added to the container image and tagged with a version, or mounted as a volume at runtime.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?

> yes, but having trouble getting the test to pass locally. I think the problem is related to the narLoader in nifi-stateless but I could use some help debugging. Everything works as expected when run from the nifi-stateless docker image

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

tpalfy commented on nifi-stateless/nifi-stateless-core/src/main/java/org/apache/nifi/stateless/core/StatelessRemoteOutputPort.java:
```diff
@@ -84,6 +87,41 @@ public StatelessRemoteOutputPort(final VersionedRemoteProcessGroup rpg, final Ve
             .build();
     }
 
+    public StatelessRemoteOutputPort(final RemoteProcessGroupDTO rpg, final RemoteProcessGroupPortDTO remotePort, final SSLContext sslContext) {
```
There isn't much logic here but maybe the checks and defaults would warrant a common private method here as well.

tpalfy commented on nifi-stateless/nifi-stateless-core/src/main/java/org/apache/nifi/stateless/runtimes/Program.java:
```diff
@@ -154,23 +161,27 @@ private static void runLocal(final String[] args, final ClassLoader systemClassL
         System.out.println("    1) " + RUN_FROM_REGISTRY + " Once --json \"{\\\"registryUrl\\\":\\\"http://172.26.198.107:61080\\\",\\\"bucketId\\\":\\\"5eec8794-01b3-4cd7-8536-0167c8b4ce8c\\\",\\\"flowId\\\": \\\"c5fa1d4f-b453-4bf5-8ff3-352352c418f3\\\"}\"");
-        System.out.println("    2) " + RUN_YARN_SERVICE_FROM_REGISTRY + " http://127.0.0.1:8088 nifi-stateless:latest kafka-to-solr 3 --file kafka-to-solr.json");
-        System.out.println("    3) " + RUN_OPENWHISK_ACTION_SERVER + " 8080");
+        System.out.println("    2) " + RUN_FROM_FLOW_XML + " Once --file /opt/nifi/nifi-stateless-configs/flow-abc.json");
```
Small inconsistency with `/Users/nifi/nifi-stateless-configs/flow-abc.json` in README.md

tpalfy commented on nifi-stateless/nifi-stateless-core/src/test/java/org/apache/nifi/stateless/core/InitFromFlowXmlIT.java:
```diff
@@ -0,0 +1,138 @@
+
+public class InitFromFlowXmlIT {
+
+//    private NiFiProperties getNiFiProperties() {
```
Whole body is commented out.

dbkegley replied:
This was failing when trying to load the processor classes from the standard processors nar. I think it has to do with the narloader but I wasn't sure where to go from there. I'll remove this and add unit tests for the flowxml parser instead

dbkegley replied:
Added FlowXml parser test

tpalfy commented on nifi-stateless/nifi-stateless-core/src/main/java/org/apache/nifi/stateless/core/StatelessFlow.java:
```diff
@@ -99,6 +122,194 @@ public StatelessFlow(final List<StatelessComponent> roots) {
         this.componentFactory = null;
     }
 
+    public StatelessFlow(final ProcessGroupDTO flow, final ExtensionManager extensionManager, final VariableRegistry variableRegistry, final List<String> failureOutputPorts,
```
Both the 2 constructors and the 2 `create...AndEnque...` methods could be normalized with a common private method.

dbkegley replied:
I'm not sure how I should approach this one. I started with using the strategy pattern to get around the sleight differences between `Versioned*` objects and `*DTO` objects but I feel like I ended up with too many generics for what I'm trying to achieve. Do you have any recommendations for how to refactor this?

dbkegley replied:
Added a new compatibility package to reconcile the differences between these classes and eliminate the code duplication

dbkegley commented:
@tpalfy Would you mind taking another look at this when you have time? I think I have addressed all of your comments and it would be great to get this merged before 1.10 is released. Thanks!

tpalfy commented on nifi-stateless/nifi-stateless-core/src/main/java/org/apache/nifi/stateless/core/StatelessFlow.java:
```diff
@@ -175,28 +194,32 @@ public StatelessFlow(final VersionedProcessGroup flow, final ExtensionManager ex
+
                     case REMOTE_INPUT_PORT:
-                        throw new IllegalArgumentException("Unsupported source type: " + source.getType());
+                        throw new IllegalArgumentException("Unsupported source type: " + source.getId());
```
Shouldn't this be `source.getType()`?

joewitt commented:
@tpalfy this needs someone to actually vet/verify as it is very large update on a relatively new codebase

@SamHjelmfelt this really looks like something you need to be reviewing and commenting on as it updates a lot of the initial stateless nifi contribution you've made.

Does it make sense for a flow.xml.gz file to be executable directly?  I'm not sure it is good to couple an internal detail of today's nifi runtime to the runtime of stateless nifi..

SamHjelmfelt commented:
I agree with Joe. Providing the option of using Stateless NiFi without the NiFi Registry does make sense, but it would be better to use NiFi templates (XML or YML) like MiNiFi does today.

Rather than using the internal configuration file of the node, users would export flows from NiFi and use them in Stateless NiFi. The biggest risk in using the flow.xml.gz directly is that it makes Stateless NiFi look like a drop in replacement for a NiFi node and encourages a usage pattern that avoids version control. The differences between Stateless NiFi and NiFi are far greater than NiFi and MiNiFi, which still does not use the flow.xml.gz.

dbkegley commented:
My goal for this was only to remove the requirement of using registry to run stateless flows so I'm in agreement about the points against using the flow.xml.gz.  Using a configuration that's more easily version controlled makes perfect sense to me, though I would rather follow MiNiFi's approach of using YAML, but that's just personal preference.

I'd still be happy to help out with this once there's a more acceptable approach. I'll keep an eye on the mailing list for those discussions. Closing this PR for now
